reader-privacy.md

```diff
@@ -0,0 +1,242 @@
+possibility of mass query snooping.
+​
+There is ongoing work on IPFS side to integrate a reader privacy technique, a.k.a, double hashing.
+Building on top of the existing approach, this document specifies how a similar technique is applied
+to IPNI in order to preserve the reader's privacy while continuing to facilitate low-latency
+provider lookup.
+​
+## Table of Contents
+
+- [Introduction](#introduction)
+- [Background](#background)
+- [Specification](#specification)
+    - [Security](#security)
+        - [Hashing and Encryption Function Upgrades](#hashing-and-encryption-function-upgrades)
+    - [Trade Offs](#trade-offs)
+    - [Threat Modelling](#threat-modelling)
+- [Implementation](#implementation)
+    - [Hashing](#hashing)
+    - [Encryption](#encryption)
+    - [Data Formats](#data-formats)
+- [Related Resources](#related-resources)
+
+For more technical implementation details please see the [Addendum](reader-privacy-addendum.md).
+​
+## Introduction
+​
+IPFS is currently lacking of many privacy protections. One of its main weak points lies in the lack 
+of privacy protections for the content routing subsystem. Currently neither readers (clients accessing files) 
+nor writers (hosts storing and distributing content) have much privacy with regard to content they publish or 
+consume. It is very easy for a content router node or a passive observer to learn which file is requested by 
+which client during the routing process, as the potential adversary easily learns about the requested `CID`. 
+A curious actor could request the same `CID` and download the associated file to monitor the user’s behavior. 
+This is obviously undesirable and has been for some time now a strong request from the community.
+
+The changes described in this specification introduce a IPNI Readres Privacy upgrade. It will prevent 
+passive observers from tracking user's actions as described above. It will also be a first step towards
+fully private IPNI protocol that will eliminate indexers as centralised observers.  
+
+### Non Goals
+
+* Writer, i.e. content provider or publisher, Privacy, which will be done in a separate specification
+* Retrieval Privacy, which is out of scope for the content routing subsystem.
+​
+## Background
+​
+Network indexers build their indexes by ingesting chains of Advertisements. Advertisement is a
+construct that allows Storage Providers to publish their CIDs in bulk (FIL deals) instead of doing
+that individually for each CID. A group of CIDs is represented by a unique ContextID as can be seen
+on the diagram below:
+​
+![Index building flow](resources/readers-privacy-1.png)
+
+## Specification
+​
+This specification improves the reader privacy by proposing changes to the Step 3, depicted above, where the client 
+supplies the content CID directly in order to lookup its corresponding providers.
+
+In order to protect the reader's privacy the proposal changes the way CID lookup works to the following:
```

Comment:
Perhaps I'm missing something obvious, why does this require making 3 round-trips instead of one? I suspect this is related to some plans you have in mind for Writer Privacy, but at first glance it seems like the server should be able to respond with all the information (peerID, multiaddrs, metadata) in a single response encrypted using the multihash.